### PR TITLE
feat(cli): manual signet fee rates, full scan cmd, disable L2 faucet, refresh -> recover, config file + more

### PR DESCRIPTION
**Pull request by Zk2u**
## Description

- Allows you to manually specify a signet fee rate with the --fee-rate option
- new “scan” command that should be used after importing a mnemonic
- featuregate L2 faucet in the CLI for the moment
- rename refresh cmd to recover
- allow configuration of bridge pubkey and network via config file
- config file location overrideable from env
- change seed length to 16 bytes from 32 (**warning** you need to reset your CLI after this PR is merged)
- update musig2 bridge pubkey

### Type of Change

-   [x] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
-   [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update
-   [x] Refactor

## Checklist

-   [x] I have performed a self-review of my code.
-   [x] I have commented my code where necessary.
-   [x] I have updated the documentation if needed.
-   [x] My changes do not introduce new warnings.
-   [ ] I have added tests that prove my changes are effective or that my feature works.
-   [x] New and existing tests pass with my changes.

## Related Issues

Closes STR-491
Closes STR-495

### CODE REVIEW

File: bin/strata-cli/src/cmd/bridge_in.rs [hunk 1]
```diff
@@ -33,13 +33,24 @@ use crate::{
 pub struct BridgeInArgs {
     #[argh(positional)]
     strata_address: Option<String>,
+
+    /// override signet fee rate in sat/vbyte. must be >1
+    #[argh(option)]
+    fee_rate: Option<String>,
```
Comment (storopoli) [inline bin/strata-cli/src/cmd/bridge_in.rs:39]: Can't this be a `usize`?

File: bin/strata-cli/src/cmd/drain.rs [hunk 2]
```diff
@@ -24,15 +24,21 @@ pub struct DrainArgs {
...
     #[argh(option, short = 's')]
     signet_address: Option<String>,
+
     /// a Strata address for Strata funds to be drained to
     #[argh(option, short = 'r')]
     strata_address: Option<String>,
+
+    /// override signet fee rate in sat/vbyte. must be >1
+    #[argh(option)]
+    fee_rate: Option<String>,
```
Comment (storopoli) [inline bin/strata-cli/src/cmd/drain.rs:34]: `usize`?

File: bin/strata-cli/src/cmd/send.rs [hunk 3]
```diff
@@ -35,13 +34,17 @@ pub struct SendArgs {
     /// address to send to
     #[argh(positional)]
     address: String,
+
+    /// override signet fee rate in sat/vbyte. must be >1
+    #[argh(option)]
+    fee_rate: Option<String>,
```
Comment (storopoli) [inline bin/strata-cli/src/cmd/send.rs:40]: `usize`?

File: bin/strata-cli/src/main.rs [hunk 4]
```diff
@@ -47,5 +47,6 @@ async fn main() {
         Commands::Receive(args) => receive(args, seed, settings, esplora).await,
         Commands::Reset(args) => reset(args, persister, settings).await,
         Commands::ChangePwd(args) => change_pwd(args, seed, persister).await,
+        Commands::Scan(args) => scan(args, seed, esplora).await,
```
Comment (storopoli) [inline bin/strata-cli/src/main.rs:50]: We also need a simple `Sync` command as well.
https://docs.rs/bdk_wallet/latest/bdk_wallet/struct.Wallet.html#method.start_sync_with_revealed_spks

Comment (storopoli) [reply bin/strata-cli/src/main.rs:50]: We can move this to a ticket and do in a separate PR if necessary

**Review by storopoli — CHANGES_REQUESTED** [review]

File: bin/strata-cli/src/main.rs (hunk 4 above)
Comment (storopoli) [reply bin/strata-cli/src/main.rs:50]: Ok! Yeah you're right in abstracting that away from the user.

File: bin/strata-cli/src/cmd/faucet.rs [hunk 5]
```diff
@@ -1,8 +1,11 @@
 use std::str::FromStr;
 
-use alloy::{primitives::Address as StrataAddress, providers::WalletProvider};
+// use alloy::{primitives::Address as StrataAddress, providers::WalletProvider};
```
Comment (storopoli) [inline bin/strata-cli/src/cmd/faucet.rs:3]: Why are we commenting this?
Do we need to remove the whole dep?
Or is this temporary? If yes, add a `FIXME` or `TODO` comment

If temporary
```suggestion
#[allow(dead_code)] // FIXME: add back when <FOO>
use alloy::{primitives::Address as StrataAddress, providers::WalletProvider};
```

File: bin/strata-cli/src/cmd/faucet.rs [hunk 6]
```diff
@@ -85,52 +86,57 @@ pub async fn faucet(args: FaucetArgs, seed: Seed, settings: Settings) {
...
+    //             None => l2w.default_signer_address(),
+    //         };
+    //         let _ = term.write_line(&format!("Claiming to Strata address {}", address));
+    //         format!(
+    //             "{base}claim_l2/{}/{}",
+    //             encode(&solution.to_le_bytes()),
+    //             address
+    //         )
+    //     }
+    // };
```
Comment (storopoli) [inline bin/strata-cli/src/cmd/faucet.rs:105]: same here

File: bin/strata-cli/src/cmd/faucet.rs [hunk 7]
```diff
@@ -12,22 +15,20 @@ use sha2::{Digest, Sha256};
...
 #[derive(FromArgs, PartialEq, Debug)]
 #[argh(subcommand, name = "faucet")]
 pub struct FaucetArgs {
-    /// either "signet" or "strata"
-    #[argh(positional)]
-    network_type: String,
-
+    // /// either "signet" or "strata"
+    // #[argh(positional)]
+    // network_type: String,
```
Comment (storopoli) [inline bin/strata-cli/src/cmd/faucet.rs:31]: same here

File: bin/strata-cli/src/cmd/faucet.rs [hunk 8]
```diff
@@ -12,22 +15,20 @@ use sha2::{Digest, Sha256};
...
 use crate::{
-    constants::NETWORK,
-    net_type::{net_type_or_exit, NetworkType},
+    // net_type::{net_type_or_exit, NetworkType},
     seed::Seed,
     settings::Settings,
-    signet::SignetWallet,
-    strata::StrataWallet,
+    signet::{print_explorer_url, SignetWallet},
+    // strata::StrataWallet,
```
Comment (storopoli) [inline bin/strata-cli/src/cmd/faucet.rs:22]: same here

**Review by storopoli — CHANGES_REQUESTED** [review]

**Review by storopoli — CHANGES_REQUESTED** [review]
As discussed in private,
let's keep the strata faucet but feature-gate it?

It might be useful in the future

File: bin/strata-cli/src/constants.rs [hunk 9]
```diff
@@ -38,9 +38,8 @@ pub const DEFAULT_ESPLORA: &str = "https://esploraapi.devnet-annapurna.stratabtc
 pub const DEFAULT_L2_HTTP_ENDPOINT: &str = "https://stratareth.devnet-annapurna.stratabtc.org";
 pub const DEFAULT_FAUCET_ENDPOINT: &str = "https://faucet.devnet-annapurna.stratabtc.org";
 pub const DEFAULT_MEMPOOL_ENDPOINT: &str = "https://mempool.devnet-annapurna.stratabtc.org";
-// FIXME: CHANGE BELOW!!!
 pub const BRIDGE_MUSIG2_PUBKEY: &str =
-    "fbd79b6b8b7fe11bad25ae89a7415221c030978de448775729c3f0a903819dd0";
+    "51acd1574ef6b1fb98ed000db6e31a7cb8d3dba302fa22e2c99a03f4cbfc2438";
```
Comment (storopoli) [inline bin/strata-cli/src/constants.rs:42]: Is this the official one?

**Review by storopoli — APPROVED** [review]
ACK 847e3c414fe1fd8c1f719bddacda563ca9a73f6c

Please **Squash Merge**

**Review by storopoli — APPROVED** [review]
ACK 3114cf7d8a4521eb980e95247715855e1b4dafe4

Squash Merge please once CI is done

File: bin/strata-cli/src/constants.rs (hunk 9 above)
Comment (sapinb) [reply bin/strata-cli/src/constants.rs:42]: @Zk2u  Please use this key:
14ced579c6a92533fa68ccc16da93b41073993cfc6cc982320645d8e9a63ee65

File: bin/strata-cli/src/constants.rs [hunk 10]
```diff
@@ -26,20 +26,20 @@ pub const PW_SALT_LEN: usize = 16;
...
 pub const AES_NONCE_LEN: usize = 12;
 /// Length of seed in bytes
-pub const SEED_LEN: usize = 32;
+pub const SEED_LEN: usize = 16;
 /// AES-256-GCM-SIV tag len
 pub const AES_TAG_LEN: usize = 16;
 
-pub const NETWORK: Network = Network::Signet;
+pub const DEFAULT_NETWORK: Network = Network::Signet;
 pub const BRIDGE_STRATA_ADDRESS: &str = "0x000000000000000000000000000000000B121d9E";
```
Comment (sapinb) [inline bin/strata-cli/src/constants.rs:34]: bridge address is now
5400000000000000000000000000000000000001

Comment (delbonis) [reply bin/strata-cli/src/constants.rs:34]: Should this be part of the params?

**Comment by storopoli** [comment]
Will JIRA pick this?
This closes STR-495

**Comment by storopoli** [comment]
I think this needs a rebase and some lint fixes

**Comment by john-light** [comment]
after running

`cargo install --git https://github.com/alpenlabs/strata strata-cli --locked --force --branch cli-feerates-and-scan`

then `strata balance signet`

I get this error:

```
strata balance signet
thread 'main' panicked at bin/strata-cli/src/main.rs:26:37:
called `Result::unwrap()` on an `Err` value: missing field `esplora`
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

running every other command returns a similar error

**Comment by delbonis** [comment]
Not sure how this is failing in clippy but nowhere else:

```
error[E0061]: this function takes 3 arguments but 2 arguments were supplied
   --> bin/strata-cli/src/cmd/faucet.rs:170:21
    |
170 |             let _ = print_explorer_url(&Txid::from_str(&body).expect("valid txid"), &term);
    |                     ^^^^^^^^^^^^^^^^^^---------------------------------------------------- argument #3 of type `&settings::Settings` is missing
    |
note: function defined here
   --> bin/strata-cli/src/signet.rs:44:8
    |
44  | pub fn print_explorer_url(txid: &Txid, term: &Term, settings: &Settings) -> Result<(), io::Error> {
    |        ^^^^^^^^^^^^^^^^^^ -----------  -----------  -------------------
help: provide the argument
    |
170 |             let _ = print_explorer_url(&Txid::from_str(&body).expect("valid txid"), &term, /* &settings::Settings */);
    |                                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: bin/strata-cli/src/seed/password.rs [hunk 11]
```diff
@@ -58,23 +54,17 @@ impl Password {
...
-                        &mut sek,
-                    )?;
-                }
-                self.seed_encryption_key = Some(sek);
-                self.seed_encryption_key(salt, version)
-            }
+    ) -> Result<[u8; 32], argon2::Error> {
+        let mut sek = [0u8; 32];
+        let (algo, ver, params) = version.params();
+        if !self.inner.is_empty() {
```
Comment (AaronFeickert) [inline bin/strata-cli/src/seed/password.rs:60]: Why check for an empty string here? A derived key from an empty string, while perhaps something of an edge case, is still valid.

Comment (AaronFeickert) [reply bin/strata-cli/src/seed/password.rs:60]: I'd pass in the empty byte array anyway, which removes the edge case entirely.

File: bin/strata-cli/src/cmd/send.rs [hunk 12]
```diff
@@ -35,25 +34,27 @@ pub struct SendArgs {
     /// address to send to
     #[argh(positional)]
     address: String,
+
+    /// override signet fee rate in sat/vbyte. must be >=1
+    #[argh(option)]
+    fee_rate: Option<u64>,
```
Comment (storopoli) [inline bin/strata-cli/src/cmd/send.rs:40]: You should add a check in https://github.com/alpenlabs/strata/blob/49b926e151bfe880a847058959d4708fc3f664cc/bin/strata-cli/src/signet.rs#L51-L59
to see if the `fee_rate` is `>=1`.

Or is this done internally in `FeeRate::from_sat_per_vb`?

File: bin/strata-cli/src/net_type.rs [hunk 13]
```diff
@@ -3,6 +3,7 @@ use std::str::FromStr;
 use console::Term;
 
 /// Represents a type of network, either Alpen's signet or Strata
+#[derive(PartialEq)]
```
Comment (storopoli) [inline bin/strata-cli/src/net_type.rs:6]:
```suggestion
#[derive(PartialEq, Eq)]
```

I think that there is a `clippy` lint that if you are deriving `Partial[Eq|Ord]` it also says that you should probably also do `[Eq|Ord]`.

File: bin/strata-cli/src/recovery.rs [hunk 14]
```diff
@@ -127,11 +128,11 @@ impl DescriptorRecovery {
     pub async fn read_descs_after_block(
         &mut self,
         height: u32,
-    ) -> Result<Vec<DescriptorTemplateOut>, OneOf<ReadDescsAfterError>> {
+    ) -> Result<Vec<(IVec, DescriptorTemplateOut)>, OneOf<ReadDescsAfterError>> {
```
Comment (storopoli) [inline bin/strata-cli/src/recovery.rs:131]: Curious: why using `IVec` here?

File: bin/strata-cli/src/seed/password.rs [hunk 15]
```diff
@@ -29,25 +30,21 @@ impl HashVersion {
...
+        let mut input = InputPassword::new().allow_empty_password(true);
         if new {
             input = input
                 .with_prompt("Create a new password (leave empty for no password, dangerous!)")
                 .with_confirmation(
                     "Confirm password (leave empty for no password, dangerous!)",
                     "Passwords didn't match",
-                )
-                .allow_empty_password(true);
+                );
```
Comment (storopoli) [inline bin/strata-cli/src/seed/password.rs:40]: Nice catch, my bad.

**Review by storopoli — CHANGES_REQUESTED** [review]
some very minor nits.

File: bin/strata-cli/src/recovery.rs (hunk 14 above)
Comment (storopoli) [reply bin/strata-cli/src/recovery.rs:131]: Thanks! TIL

**Review by AaronFeickert — APPROVED** [review]
Reviewed for passphrase and seed encryption. LGTM.

**Review by storopoli — APPROVED** [review]
ACK 630a907a5a2eb0d66b99dfb2e1b137f843dbbb17

Nice work!

**Comment by storopoli** [comment]
We could remove the codecov CI job from the requirements to merge. What do you think @delbonis. Then, it will be easier to merge PRs and not depend on your admin rights.

PS: If you are merging this, please **squash and merge**?

**Pull request merged**